service/tablet_allocator.cc

```diff
@@ -922,7 +922,31 @@ class load_balancer {
                 return bool(t.transition);
             };
 
-            auto first_non_matching_replicas = [] (tablet_replica_set r1, tablet_replica_set r2) -> std::optional<std::pair<tablet_replica, tablet_replica>> {
+            // Enable rack-aware co-location when there's more than one rack. But in practice, merge
+            // will only be able to proceed when RF=#racks, since in this configuration, the amount
+            // of cross-rack migrations will be ZERO.
+            // When RF!=#racks, it's not always possible to achieve ZERO cross-rack migrations,
+            // which is crucial for not lowering availability and not affecting pairing consistency.
+            auto rack_aware_enabled = std::invoke([&] () {
+                auto& topology = _tm->get_topology();
+                auto [_, rs] = get_schema_and_rs(table);
+                auto dc_rf = rs->get_replication_factor(dc);
+                const auto& racks = topology.get_datacenter_rack_nodes().at(dc);
+                auto multi_rack = racks.size() > 1;
+                if (multi_rack && dc_rf != racks.size()) {
+                    lblogger.warn("Table {} has RF different than number of racks in DC {}, so merge can get stuck. RF={}, racks={}",
```

Comment:
We need a better way of signaling such conditions. This log can be missed because it happens only when there is need for merge which is asynchronous with admin action. And when it happens, it will spam the logs constantly. 

It would be good to integrate into alerting system so that administrator can take action before it's too late, and make logging edge-triggered rather than level-triggered.

Maybe not for this PR.

Reply:
This is irrelevant if we make this a per-table property and maintain invariants in the tablet allocator.